### PR TITLE
Dockle parser: switch to hash_code based deduplication

### LINKED ISSUE #4788: Dockle Scan re import not matching same vulnerabilities
**Bug description**
When **re**importing Dockle scans, the vulnerabilities of the new report do not replace the current ones when they are identical. They are not even seen as duplicates.

**Steps to reproduce**
1. Execute a first Docker Scan and upload the report to DefectDojo
`curl -X POST "${DD_URL}/api/v2/import-scan/"
      -H "accept: application/json"
      -H "Authorization: Token ${DD_TOKEN}"
      -H "Content-Type: multipart/form-data"
      -F "scan_date=${DATE}"
      -F "minimum_severity=Info"
      -F "active=true"
      -F "verified=false"
      -F "scan_type=Dockle Scan"
      -F "engagement=${ENGAGEMENT_ID}"
      -F "file=@/${CI_PROJECT_DIR}/report_dockle.json"
      -F "close_old_findings=false"
      -F "DEFECTDOJO_PUSH_TO_JIRA=${DEFECTDOJO_PUSH_TO_JIRA}"
      -F "environment=${CI_ENVIRONMENT_NAME}"
      -F "version=${VERSION_RELEASE}"
      -F "branch_tag=${CI_COMMIT_REF_NAME}"
      -F "commit_hash=${CI_COMMIT_SHORT_SHA}"`

2. Execute another scan and reimport this report to that same Test.
`curl -X POST "${DD_URL}/api/v2/reimport-scan/"
      -H "accept: application/json"
      -H "Authorization: Token ${DD_TOKEN}"
      -H "Content-Type: multipart/form-data"
      -F "scan_date=${DATE}"
      -F "minimum_severity=Info"
      -F "active=true"
      -F "verified=false"
      -F "test=${TEST_ID}"
      -F "scan_type=Dockle Scan"
      -F "file=@/${CI_PROJECT_DIR}/report_dockle.json"
      -F "close_old_findings=false"
      -F "DEFECTDOJO_PUSH_TO_JIRA=${DEFECTDOJO_PUSH_TO_JIRA}"
      -F "version=${VERSION_RELEASE}"
      -F "branch_tag=${CI_COMMIT_REF_NAME}"
      -F "commit_hash=${CI_COMMIT_SHORT_SHA}"`
3. You should see your vulnerabilities twice but not marked as duplicates

**Expected behavior**
Do not add vulnerabilities that are the same as new findings, or at least detect that they are duplicates.

**Deployment method**
- [ ] Docker
- [X] Kubernetes
- [ ] GoDojo
- [ ] setup.bash / legacy-setup.bash

**Environment information**
 - Operating System: Ubuntu

**Screenshots** (optional)
First import
First import
Then reimport
Then reimport

**Additional context** (optional)
I have had the same issue with SonarQube re imports, but Sonar reimport does identify duplicates and hence I can configure DD to automatically delete them

I also wonder what is `close_old_findings` expected behaviour when using **re**import

### PR DESCRIPTION
**Pull request by SoaAlex**
Fixes #4788

Default legacy dedupe algo did not work properly for Dockle has some findings were still not marked as duplicates even if they are the exact sale.
Also fixed description sorting

Hash code fileds are title, description, and vuln_id_from_tool

### CODE REVIEW

**Comment by SoaAlex** [comment]
Please wait before merging this because the parser seems to not order the description in the same way when parsing even thought it's the same vulnerability (so they don't have the same hash_code and are not deduped)
image
image

I will try to fix it as well

**Comment by SoaAlex** [comment]
We have 2 solutions.
Consider only the title + vuln_id_from_tool

Or make this line always output the same description:
`description = "\n".join(item.get('alerts', []))`

**Comment by SoaAlex** [comment]
Nevermind, the issue is not from the parser.py but from the way how Dockle generates the description.
It randomly sorts the description...

Therefore, the only way is generating the hash code from vuln_id_from_tool and from title only

**Comment by StefanFl** [comment]
Having the description might be good, if the list of files changes. On the other hand an older finding would have to be closed manually in that case. I tend to the option of keeping the description as part of the hashcode and to sort the list alphabetically. Should be done with `description = "\n".join(sort(item.get('alerts', [])))` or so (haven't tested it).

**Comment by SoaAlex** [comment]
Sorting alphabetically is a good idea, I agree with you.

I will do this change tomorrow then, I have to go for now.

**Comment by SoaAlex** [comment]
Made the change quickly. Should be fine.
Will do further testing tomorrow

**Review by StefanFl — APPROVED** [review]

**Comment by SoaAlex** [comment]
It's works fine. I've tested it locally.

**Comment by StefanFl** [comment]
Great job @SoaAlex. The failing checks are not related to your PR, we are working on them.

**Review by StefanFl — APPROVED** [review]

File: dojo/tools/dockle/parser.py
Comment (damiencarol) [inline dojo/tools/dockle/parser.py:42]:
```suggestion
            description = "\n".join(item.get('alerts', []).sort())
```

**Review by damiencarol — DISMISSED** [review]

**Comment by damiencarol** [comment]
@SoaAlex will accept and merge once the K8s pb is fixed

File: dojo/tools/dockle/parser.py
Comment (Maffooch) [reply dojo/tools/dockle/parser.py:42]: Good suggestion here

Comment (SoaAlex) [reply dojo/tools/dockle/parser.py:42]: I've had already tried this way but it doesn't work because the .sort() function returns `NoneType`.
You must have 3 separate ugly lines but at least it works.
image

Comment (SoaAlex) [reply dojo/tools/dockle/parser.py:42]: I've had already tried this way but it doesn't work because the .sort() function returns NoneType.
You must have 3 separate ugly lines but at least it works.
Capture d’écran 2021-07-30 à 09 41 34
Capture d’écran 2021-07-30 à 09 36 37

**Comment by SoaAlex** [comment]
Thank you all

**Comment by SoaAlex** [comment]
Is the k8s problem fixed ?

**Comment by StefanFl** [comment]
@SoaAlex  Yes it is, can you please rebase your branch against dev and push it? Then all the tests should get green

**Comment by SoaAlex** [comment]
It's just `git rebase dev` right ?

**Comment by SoaAlex** [comment]
Okay, somehow managed to rebase again without including all previous commits.

**Review by Maffooch — APPROVED** [review]

**Comment by SoaAlex** [comment]
So can this PR be merged ?

**Comment by StefanFl** [comment]
We still need @damiencarol's approval for the changes

**Pull request merged**
